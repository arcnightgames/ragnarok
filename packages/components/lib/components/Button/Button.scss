@use "sass:math";
@import "@arcnight/styles/scss/lib";

/*
 * Button shape
 */
.button {
  --padding-inline: #{em(15)};

  display: inline-block;
  padding-block: em(7);
  padding-inline: var(--padding-inline);
  border: 1px solid var(--button-default-border-color, transparent);
  border-radius: var(--border-radius-md);
  outline: none;
  background-color: var(--button-default-background-color, transparent);
  color: var(--button-default-text-color, inherit);
  text-align: center;
  cursor: pointer;
  transition: 0.15s easing(ease-out-quint);

  @supports selector(:focus-visible:has(a, b)) {
    // had to do this because of Safari
    outline: unset;
  }

  &,
  &:hover,
  &:focus,
  &:focus-within,
  &:active {
    text-decoration: none;
  }

  &:hover,
  &.is-hover {
    border-color: var(--button-hover-border-color, transparent);
    background-color: var(--button-hover-background-color, transparent);
    color: var(--button-hover-text-color, inherit);
    opacity: 1;
  }

  @supports not selector(:focus-visible:has(a, b)) {
    &:focus,
    &:focus-within {
      border-color: var(--focus-outline);
      box-shadow: 0 0 0 #{$focus-ring-size} var(--focus-ring);
      color: var(--button-default-text-color, inherit);
    }
  }

  @supports selector(:focus-visible:has(a, b)) {
    &:focus {
      color: var(--button-default-text-color, inherit);
    }

    &:focus-visible,
    &:has(*:focus-visible) {
      border-color: var(--focus-outline);
      box-shadow: 0 0 0 #{$focus-ring-size} var(--focus-ring);
      color: var(--button-default-text-color, inherit);
    }
  }

  &:active,
  &.is-active,
  &[aria-expanded="true"]:not([aria-controls]) {
    border-color: var(--button-active-border-color, transparent);
    background-color: var(--button-active-background-color, transparent);
    color: var(--button-active-text-color, inherit);
    opacity: 1;
  }

  &[disabled],
  &.is-disabled {
    pointer-events: none;
    background-color: var(--button-hover-background-color, transparent);

    &:not([aria-busy="true"]) {
      border-color: transparent;
      background-color: var(--weak, transparent);
      color: var(--text-weak, inherit);
      opacity: 0.5;

      &[class*="button-ghost"] {
        background-color: var(--button-default-background-color, transparent);
      }
    }
  }
}

/*
 * Button variants
 */
@each $color in (primary, secondary, danger, warning, success, info) {
  // Solid shape
  .button-solid-#{$color} {
    $solid-text-color: var(--#{if($color == "primary" or $color == "secondary", "text-norm", dash-join($color, "contrast"))});

    --button-default-background-color: var(--#{$color});
    --button-hover-background-color: var(--#{dash-join($color, "major-1")});
    --button-active-background-color: var(--#{dash-join($color, "major-2")});
    --button-default-text-color: #{$solid-text-color};
    --button-hover-text-color: #{$solid-text-color};
    --button-active-text-color: #{$solid-text-color};

    @if $color == "primary" {
      --focus-outline: var(--norm-major-2);
    }
  }

  // Outline shape
  .button-outline-#{$color} {
    --button-default-background-color: var(--background-norm);
    --button-hover-background-color: var(--#{dash-join($color, "minor-2")});
    --button-active-background-color: var(--#{dash-join($color, "minor-1")});
    --button-default-border-color: var(--#{dash-join($color, "major-1")});
    --button-hover-border-color: var(--#{dash-join($color, "major-2")});
    --button-active-border-color: var(--#{dash-join($color, "major-3")});
    --button-default-text-color: var(
      --#{if($color == "primary" or $color == "secondary", "text-norm", dash-join($color, "major-1"))}
    );
    --button-hover-text-color: var(
      --#{if($color == "primary" or $color == "secondary", "text-norm", dash-join($color, "major-2"))}
    );
    --button-active-text-color: var(
      --#{if($color == "primary" or $color == "secondary", "text-norm", dash-join($color, "major-3"))}
    );
  }

  // Ghost shape
  .button-ghost-#{$color} {
    --button-default-background-color: var(--interaction-default);
    --button-hover-background-color: var(--interaction-default-hover);
    --button-active-background-color: var(--interaction-default-active);
    --button-default-text-color: var(--#{if($color == "primary" or $color == "secondary", "text-norm", dash-join($color, "major-1"))});
    --button-hover-text-color: var(--#{if($color == "primary" or $color == "secondary", "text-norm", dash-join($color, "major-2"))});
    --button-active-text-color: var(--#{if($color == "primary" or $color == "secondary", "text-norm", dash-join($color, "major-3"))});
  }
}

/*
 * Modifiers
 */
.button-small {
  --padding-inline: #{em(11)};

  padding-block: em(3);
}

.button-large {
  --padding-inline: #{em(19, 16)};

  padding-block: em(9, 16);
  font-size: em(16);
  line-height: #{math.div(24, 16)};
}

// Loading state
.button[aria-busy="true"] {
  position: relative;
  padding-inline-end: em(40);

  &.button-small {
    padding-inline-end: em(32);
  }

  &.button-large {
    padding-inline-end: em(44);
  }

  .button-loader-container {
    inset-inline-end: em(16);
  }

  &.button-small .button-loader-container {
    inset-inline-end: em(10);
  }

  &.button-large .button-loader-container {
    inset-inline-end: em(20);
  }
}

.button-underline[aria-busy="true"] {
  position: relative;
  padding-inline-end: em(20);

  .button-loader-container {
    inset-inline-end: 0;
  }
}

.button-loader-container {
  position: absolute;
  inset-block: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  animation: 0.25s ease-out both anime-button-loader-container-fade-in;

  @keyframes anime-button-loader-container-fade-in {
    from {
      opacity: 0;
    }

    to {
      opacity: 1;
    }
  }
}

.w-full {
  width: 100%;
}
