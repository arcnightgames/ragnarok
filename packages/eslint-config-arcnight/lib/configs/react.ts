import {
  pluginJSX,
  pluginReact,
  PluginReactHooks,
  pluginTestingLibrary,
} from "../plugins";
import {ConfigItem} from "../types";

export const react = (): ConfigItem[] => {
  return [
    {
      name: "arcnight:react",
      plugins: {
        jsx: pluginJSX,
        react: pluginReact,
        "react-hooks": PluginReactHooks,
        "testing-library": pluginTestingLibrary,
      },
    },
    {
      rules: {
        "no-underscore-dangle": [
          "error",
          {
            allow: ["__REDUX_DEVTOOLS_EXTENSION_COMPOSE__"],
            allowAfterThis: false,
            allowAfterSuper: false,
            enforceInMethodNames: true,
          },
        ],
        "class-methods-use-this": [
          "error",
          {
            exceptMethods: [
              "render",
              "getInitialState",
              "getDefaultProps",
              "getChildContext",
              "componentWillMount",
              "UNSAFE_componentWillMount",
              "componentDidMount",
              "componentWillReceiveProps",
              "UNSAFE_componentWillReceiveProps",
              "shouldComponentUpdate",
              "componentWillUpdate",
              "UNSAFE_componentWillUpdate",
              "componentDidUpdate",
              "componentWillUnmount",
              "componentDidCatch",
              "getSnapshotBeforeUpdate",
            ],
          },
        ],
        "react/display-name": ["off", {ignoreTranspilerName: false}],
        "react/forbid-prop-types": [
          "error",
          {
            forbid: ["any", "array", "object"],
            checkContextTypes: true,
            checkChildContextTypes: true,
          },
        ],
        "react/forbid-dom-props": ["off", {forbid: []}],
        "react/jsx-boolean-value": ["error", "never", {always: []}],
        "react/jsx-handler-names": [
          "off",
          {
            eventHandlerPrefix: "handle",
            eventHandlerPropPrefix: "on",
          },
        ],
        "react/jsx-key": "off",
        "react/jsx-no-bind": [
          "error",
          {
            ignoreRefs: true,
            allowArrowFunctions: true,
            allowFunctions: false,
            allowBind: false,
            ignoreDOMComponents: true,
          },
        ],
        "react/jsx-no-duplicate-props": ["error", {ignoreCase: true}],
        "react/jsx-no-literals": ["off", {noStrings: true}],
        "react/jsx-no-undef": "error",
        "react/jsx-pascal-case": [
          "error",
          {
            allowAllCaps: true,
            ignore: [],
          },
        ],
        "react/sort-prop-types": [
          "off",
          {
            ignoreCase: true,
            callbacksLast: false,
            requiredFirst: false,
            sortShapeProp: true,
          },
        ],
        "react/jsx-sort-prop-types": "off",
        "react/jsx-sort-props": [
          "off",
          {
            ignoreCase: true,
            callbacksLast: false,
            shorthandFirst: false,
            shorthandLast: false,
            noSortAlphabetically: false,
            reservedFirst: true,
          },
        ],
        "react/jsx-uses-react": ["error"],
        "react/jsx-uses-vars": "error",
        "react/no-danger": "warn",
        "react/no-deprecated": ["error"],
        "react/no-did-mount-set-state": "off",
        "react/no-did-update-set-state": "error",
        "react/no-will-update-set-state": "error",
        "react/no-direct-mutation-state": "off",
        "react/no-is-mounted": "error",
        "react/no-multi-comp": "off",
        "react/no-set-state": "off",
        "react/no-string-refs": "error",
        "react/no-unknown-property": "error",
        "react/prefer-es6-class": ["error", "always"],
        "react/prefer-stateless-function": [
          "error",
          {ignorePureComponents: true},
        ],
        "react/require-render-return": "error",
        "react/self-closing-comp": "error",
        "react/sort-comp": [
          "error",
          {
            order: [
              "static-variables",
              "static-methods",
              "instance-variables",
              "lifecycle",
              "/^handle.+$/",
              "/^on.+$/",
              "getters",
              "setters",
              "/^(get|set)(?!(InitialState$|DefaultProps$|ChildContext$)).+$/",
              "instance-methods",
              "everything-else",
              "rendering",
            ],
            groups: {
              lifecycle: [
                "displayName",
                "propTypes",
                "contextTypes",
                "childContextTypes",
                "mixins",
                "statics",
                "defaultProps",
                "constructor",
                "getDefaultProps",
                "getInitialState",
                "state",
                "getChildContext",
                "getDerivedStateFromProps",
                "componentWillMount",
                "UNSAFE_componentWillMount",
                "componentDidMount",
                "componentWillReceiveProps",
                "UNSAFE_componentWillReceiveProps",
                "shouldComponentUpdate",
                "componentWillUpdate",
                "UNSAFE_componentWillUpdate",
                "getSnapshotBeforeUpdate",
                "componentDidUpdate",
                "componentDidCatch",
                "componentWillUnmount",
              ],
              rendering: ["/^render.+$/", "render"],
            },
          },
        ],
        "react/jsx-no-target-blank": ["error", {enforceDynamicLinks: "always"}],
        "react/jsx-no-comment-textnodes": "error",
        "react/no-render-return-value": "error",
        "react/require-optimization": ["off", {allowDecorators: []}],
        "react/no-find-dom-node": "error",
        "react/forbid-elements": ["off", {forbid: []}],
        "react/no-danger-with-children": "error",
        "react/no-unused-prop-types": [
          "error",
          {
            customValidators: [],
            skipShapeProps: true,
          },
        ],
        "react/style-prop-object": "error",
        "react/no-unescaped-entities": "error",
        "react/no-children-prop": "error",
        "react/no-array-index-key": "error",
        "react/forbid-foreign-prop-types": ["warn", {allowInPropTypes: true}],
        "react/void-dom-elements-no-children": "error",
        "react/default-props-match-prop-types": [
          "error",
          {allowRequiredDefaults: false},
        ],
        "react/no-redundant-should-component-update": "error",
        "react/no-unused-state": "error",
        "react/boolean-prop-naming": [
          "off",
          {
            propTypeNames: ["bool", "mutuallyExclusiveTrueProps"],
            rule: "^(is|has)[A-Z]([A-Za-z0-9]?)+",
            message: "",
          },
        ],
        "react/no-typos": "error",
        "react/jsx-curly-brace-presence": [
          "error",
          {props: "never", children: "never"},
        ],
        "react/destructuring-assignment": ["error", "always"],
        "react/no-access-state-in-setstate": "error",
        "react/button-has-type": [
          "error",
          {
            button: true,
            submit: true,
            reset: false,
          },
        ],
        "react/jsx-child-element-spacing": "off",
        "react/no-this-in-sfc": "error",
        "react/jsx-max-depth": "off",
        "react/no-unsafe": "off",
        "react/jsx-fragments": ["error", "syntax"],
        "react/state-in-constructor": ["error", "always"],
        "react/static-property-placement": ["error", "property assignment"],
        "react/prefer-read-only-props": "off",
        "react/jsx-no-script-url": [
          "error",
          [
            {
              name: "Link",
              props: ["to"],
            },
          ],
        ],
        "react/jsx-no-useless-fragment": "error",
        "react/no-adjacent-inline-elements": "off",
        "react/function-component-definition": [
          "error",
          {
            namedComponents: ["arrow-function"],
          },
        ],
        "react/jsx-newline": "off",
        "react/jsx-no-constructed-context-values": "error",
        "react/no-unstable-nested-components": "error",
        "react/no-namespace": "error",
        "react/prefer-exact-props": "error",
        "react/no-arrow-function-lifecycle": "error",
        "react/no-invalid-html-attribute": "error",
        "react/no-unused-class-component-methods": "error",
        "react/hook-use-state": "off",
        "react/iframe-missing-sandbox": "off",
        "react/jsx-no-leaked-render": "off",
        "react/no-object-type-as-default-prop": "off",
        "react/sort-default-props": [
          "off",
          {
            ignoreCase: false,
          },
        ],
        "react/forbid-component-props": [
          "warn",
          {
            forbid: [
              {
                propName: "data-test-id",
                message: "Please use `data-testid` instead",
              },
            ],
          },
        ],
        "react/jsx-filename-extension": [1, {extensions: [".jsx", ".tsx"]}],
        "react/jsx-props-no-spreading": "off",
        "react/prop-types": "warn",
        "react/react-in-jsx-scope": "off",
        "react/require-default-props": "off",
        "react-hooks/rules-of-hooks": "error",
        "jsx/alt-text": "error",
        "jsx/anchor-has-content": "error",
        "jsx/aria-activedescendant-has-tabindex": "error",
        "jsx/aria-props": "error",
        "jsx/aria-proptypes": "error",
        "jsx/aria-role": "error",
        "jsx/aria-unsupported-elements": "error",
        "jsx/autocomplete-valid": "error",
        "jsx/heading-has-content": "error",
        "jsx/html-has-lang": "error",
        "jsx/iframe-has-title": "error",
        "jsx/no-access-key": "error",
        "jsx/no-distracting-elements": "error",
        "jsx/no-noninteractive-element-to-interactive-role": [
          "error",
          {
            ul: [
              "listbox",
              "menu",
              "menubar",
              "radiogroup",
              "tablist",
              "tree",
              "treegrid",
            ],
            ol: [
              "listbox",
              "menu",
              "menubar",
              "radiogroup",
              "tablist",
              "tree",
              "treegrid",
            ],
            li: ["menuitem", "option", "row", "tab", "treeitem"],
            table: ["grid"],
            td: ["gridcell"],
            fieldset: ["radiogroup", "presentation"],
          },
        ],
        "jsx/no-redundant-roles": "error",
        "jsx/role-has-required-aria-props": "error",
        "jsx/role-supports-aria-props": "error",
        "jsx/scope": "error",
        "jsx/tabindex-no-positive": "error",
        "jsx/anchor-ambiguous-text": "warn",
        "jsx/anchor-is-valid": "warn",
        "jsx/control-has-associated-label": "warn",
        "jsx/img-redundant-alt": "off",
        "jsx/interactive-supports-focus": "warn",
        "jsx/label-has-associated-control": "warn",
        "jsx/media-has-caption": "warn",
        "jsx/mouse-events-have-key-events": "warn",
        "jsx/no-aria-hidden-on-focusable": "warn",
        "jsx/no-autofocus": "off",
        "jsx/no-interactive-element-to-noninteractive-role": "warn",
        "jsx/no-noninteractive-tabindex": "warn",
        "jsx/no-static-element-interactions": "warn",
      },
      settings: {
        "import/resolver": {
          node: {
            extensions: [".js", ".jsx", ".json"],
          },
        },
        react: {
          version: "detect",
        },
        propWrapperFunctions: ["forbidExtraProps", "exact", "Object.freeze"],
      },
    },
  ];
};
