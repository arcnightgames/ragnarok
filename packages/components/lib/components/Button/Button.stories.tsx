import {Meta} from "@storybook/react";

import Button from "./Button";

export default {
  component: Button,
  title: "components/Button",
} as Meta<typeof Button>;

export const Playground = ({...args}) => <Button {...args}>test</Button>;

Playground.argTypes = {
  size: {
    options: ["small", "medium", "large"],
    control: {
      type: "select",
      defaultValue: "medium",
    },
  },
  color: {
    options: ["primary", "secondary", "danger", "warning", "success"],
    control: {
      type: "select",
      defaultValue: "primary",
    },
  },
  shape: {
    options: ["solid", "ghost", "outline"],
    control: {
      type: "select",
      defaultValue: "solid",
    },
  },
  loading: {
    control: {
      type: "boolean",
    },
  },
  fullWidth: {
    control: {
      type: "boolean",
    },
  },
};
