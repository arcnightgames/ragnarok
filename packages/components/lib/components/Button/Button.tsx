import "./Button.scss";

import {PolymorphicPropsWithRef} from "@arcnight/react-polymorphic-types";
import {ThemeColor, ThemeColorUnion} from "@arcnight/shared";
import {clsx} from "@arcnight/utils";
import {ElementType, ForwardedRef, forwardRef} from "react";

import {CircleLoader} from "../CircleLoader";

export type ButtonShape = "ghost" | "solid" | "outline";

export type ButtonSize = "small" | "medium" | "large";

interface ButtonOwnProps {
  /**
   * Whether the button should render a loader.
   * Button is disabled when this prop is true.
   */
  loading?: boolean;
  /**
   * Controls the shape of the button.
   * - `ghost` for texted button
   * - `solid` for filled button
   * - `outline` for bordered button
   */
  shape?: ButtonShape;
  /**
   * Controls the colors of the button.
   * Exact styles applied to depend on the chosen shape as well.
   */
  color?: ThemeColorUnion;
  /**
   * Controls how large the button should be.
   */
  size?: ButtonSize;
  /**
   * Controls the width of the button.
   * - `false` for width of text length
   * - `true` for width of parent container
   */
  fullWidth?: boolean;
  /**
   * Puts the button in a disabled state.
   */
  disabled?: boolean;
  /**
   * Locator for e2e tests.
   */
  "data-testid"?: string;
}

export type ButtonProps<E extends ElementType> = PolymorphicPropsWithRef<
  ButtonOwnProps,
  E
>;

const defaultElement = "button";

const Button = <E extends ElementType = typeof defaultElement>(
  {
    loading = false,
    disabled = false,
    fullWidth = false,
    className,
    tabIndex,
    children,
    shape = "solid",
    color = ThemeColor.Primary,
    size = "medium",
    as,
    "data-testid": dataTestId,
    ...restProps
  }: ButtonProps<E>,
  ref: ForwardedRef<Element>,
) => {
  const isDisabled = loading || disabled;

  const Element: ElementType = as || defaultElement;

  const buttonClassName = clsx(
    "button",
    fullWidth && "w-full",
    size !== "medium" && `button-${size}`,
    `button-${shape}-${color}`,
    Element !== "button" && "inline-block text-center",
    className,
  );

  const roleProps =
    restProps.onClick && !restProps.type ? {role: "button"} : undefined;

  return (
    <Element
      ref={ref}
      className={buttonClassName}
      disabled={isDisabled}
      tabIndex={isDisabled ? -1 : tabIndex}
      aria-busy={loading}
      data-testid={dataTestId}
      {...roleProps}
      {...restProps}
    >
      {children}
      {loading && (
        <span className="button-loader-container">
          <CircleLoader />
        </span>
      )}
    </Element>
  );
};

export default forwardRef(Button);
