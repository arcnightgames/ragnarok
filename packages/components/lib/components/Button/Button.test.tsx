import "@testing-library/jest-dom";

import {fireEvent, render, screen} from "@testing-library/react";

import Button from "./Button";

describe("Button Component", () => {
  it("renders button with default props", () => {
    render(<Button>Click me</Button>);
    const button = screen.getByRole("button");
    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent("Click me");
  });

  it("renders button with loading state", () => {
    render(<Button loading>Click me</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveAttribute("aria-busy", "true");
  });

  it("handles button click", () => {
    const handleClick = vi.fn();
    render(<Button onClick={handleClick}>Click me</Button>);
    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalled();
  });

  it("renders disabled button", () => {
    render(<Button disabled>Click me</Button>);
    const button = screen.getByRole("button");
    expect(button).toBeDisabled();
  });

  it("renders button with custom class", () => {
    render(<Button className="custom-class">Click me</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("custom-class");
  });

  it("renders button with custom size", () => {
    render(<Button size="large">Click me</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("button-large");
  });

  it("renders button with custom shape and color", () => {
    render(
      <Button shape="ghost" color="secondary">
        Click me
      </Button>,
    );
    const button = screen.getByRole("button");
    expect(button).toHaveClass("button-ghost-secondary");
  });
});
