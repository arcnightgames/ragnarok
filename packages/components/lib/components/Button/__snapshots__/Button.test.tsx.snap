// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > renders with custom color 1`] = `
<button
  aria-busy="false"
  class="button button-solid-secondary"
>
  Click me
</button>
`;

exports[`Button Component > renders with custom data-testid 1`] = `
<button
  aria-busy="false"
  class="button button-solid-primary"
  data-testid="test-button"
>
  Click me
</button>
`;

exports[`Button Component > renders with custom shape 1`] = `
<button
  aria-busy="false"
  class="button button-outline-primary"
>
  Click me
</button>
`;

exports[`Button Component > renders with custom size 1`] = `
<button
  aria-busy="false"
  class="button button-small button-solid-primary"
>
  Click me
</button>
`;

exports[`Button Component > renders with default props 1`] = `
<button
  aria-busy="false"
  class="button button-solid-primary"
>
  Click me
</button>
`;

exports[`Button Component > renders with disabled state 1`] = `
<button
  aria-busy="false"
  class="button button-solid-primary"
  disabled=""
  tabindex="-1"
>
  Click me
</button>
`;

exports[`Button Component > renders with full width 1`] = `
<button
  aria-busy="false"
  class="button w-full button-solid-primary"
>
  Click me
</button>
`;

exports[`Button Component > renders with loading state 1`] = `
<button
  aria-busy="true"
  class="button button-solid-primary"
  disabled=""
  tabindex="-1"
>
  Click me
  <span
    class="button-loader-container"
  >
    <svg
      class="circle-loader"
      viewBox="0 0 16 16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <circle
          cx="8"
          cy="8"
          id="circle_loader_1"
          r="7"
        />
      </defs>
      <use
        class="circle-loader-track"
        href="#circle_loader_1"
      />
      <use
        class="circle-loader-circle"
        href="#circle_loader_1"
      />
    </svg>
  </span>
</button>
`;
